File: src/SmithWaterman.FSharp/Algorithm.fs
module SmithWaterman.FSharp.Algorithm

open System.Text

module private Helper =

    let createMatrix (subjectLength, queryLength) =
        Array.zeroCreate ((subjectLength + 1) * (queryLength + 1))

    let inline idx columnWidth (i, j) = i * columnWidth + j

    let inline getMatrix columnWidth (matrix: array<'a>) (i, j) = matrix[idx columnWidth (i, j)]

    let inline setMatrix columnWidth (matrix: array<'a>) (i, j) value = matrix[idx columnWidth (i, j)] <- value

    let calculateMatrix
        (subject: string, query: string)
        (matrix: array<int>)
        (matchBonus, mismatchPenalty, gapPenalty)
        =
        let subjectLength = subject.Length
        let queryLength = query.Length

        let getMatrix = getMatrix (queryLength + 1) matrix
        let setMatrix = setMatrix (queryLength + 1) matrix

        for i in 1..subjectLength do
            for j in 1..queryLength do
                let diagonal = getMatrix (i - 1, j - 1)
                let up = getMatrix (i - 1, j)
                let left = getMatrix (i, j - 1)

                let matching =
                    if subject[i - 1] = query[j - 1] then
                        diagonal + matchBonus
                    else
                        diagonal - mismatchPenalty

                let upwardGap = up - gapPenalty
                let leftwardGap = left - gapPenalty
                setMatrix (i, j) (Seq.max [ matching; upwardGap; leftwardGap; 0 ])

        ()

    type Direction =
        | Upward
        | Leftward
        | Diagonal
        | Stop

    let getDirection (getMatrix: (int * int) -> int) (i, j) (query: string, subject: string) (matchBonus, mismatchPenalty, gapPenalty)=
        let cur = getMatrix (i, j)

        let diag = getMatrix (i - 1, j - 1)
        let diagMatch = cur - matchBonus
        let diagMismatch = cur + mismatchPenalty

        let up = getMatrix (i - 1, j)
        let upGap = cur + gapPenalty

        let left = getMatrix (i, j - 1)
        let leftGap = cur + gapPenalty

        if cur = 0 then Stop
        elif diag = diagMatch && subject[i - 1] = query[j - 1] then Diagonal
        elif diag = diagMismatch && subject[i - 1] <> query[j - 1] then Diagonal
        elif up = upGap then Upward
        elif left = leftGap then Leftward
        else Stop

    let backtrackAlignment (subject: string, query: string) (matrix: array<int>) scoreProperty=
        let subjectLength = subject.Length
        let queryLength = query.Length

        let maxScoreIndex =
            let mutable bestV = matrix[0]
            let mutable bestI = 0

            for i in 0 .. matrix.Length - 1 do
                if matrix[i] > bestV then
                    bestV <- matrix[i]
                    bestI <- i

            bestI

        let subjectLine = new StringBuilder(max subjectLength queryLength)
        let queryLine = new StringBuilder(max subjectLength queryLength)

        // this is a branched tail-call and .NET JIT should be smart enough to optimize it
        let rec backtrackRecursive (currentRow, currentCol) scoreProperty =
            if currentRow = 0 || currentCol = 0 then
                ()
            else
                let getMatrix = getMatrix (queryLength + 1) matrix

                let direction = getDirection getMatrix (currentRow, currentCol) (query, subject) scoreProperty

                match direction with
                | Upward ->
                    let _ = subjectLine.Insert(0, subject[currentRow - 1])
                    let _ = queryLine.Insert(0, '-')
                    backtrackRecursive (currentRow - 1, currentCol) scoreProperty
                | Leftward ->
                    let _ = subjectLine.Insert(0, '-')
                    let _ = queryLine.Insert(0, query[currentCol - 1])
                    backtrackRecursive (currentRow, currentCol - 1) scoreProperty
                | Diagonal ->
                    let _ = subjectLine.Insert(0, subject[currentRow - 1])
                    let _ = queryLine.Insert(0, query[currentCol - 1])
                    backtrackRecursive (currentRow - 1, currentCol - 1) scoreProperty
                | Stop -> ()

        backtrackRecursive (maxScoreIndex / (queryLength + 1), maxScoreIndex % (queryLength + 1)) scoreProperty

        subjectLine.ToString(), queryLine.ToString()

let smithWaterman (subject: string, query: string) (matchBonus, mismatchPenalty, gapPenalty) =
    let subjectLength = subject.Length
    let queryLength = query.Length

    let matrix = Helper.createMatrix (subjectLength, queryLength)

    Helper.calculateMatrix (subject, query) matrix (matchBonus, mismatchPenalty, gapPenalty)

    Helper.backtrackAlignment (subject, query) matrix (matchBonus, mismatchPenalty, gapPenalty)

File: scala/src/SmithWaterman/Algorithm.scala
package SmithWaterman

case class ScoreInfo(matchBonus: Int, mismatchPenalty: Int, gapPenalty: Int)

class Algorithm(subject: String, query: String, scoreInfo: ScoreInfo):
  private var matrix = Array.fill((subject.length + 1) * (query.length + 1))(0)
  private val colWidth = subject.length + 1

  private def getMatrix(row: Int, col: Int): Int = matrix(row * colWidth + col)
  private def setMatrix(row: Int, col: Int, value: Int): Unit =
    matrix = matrix.updated(row * colWidth + col, value)

  private def calculateMatrix(): Unit =
    for (row <- 1 to subject.length)
      for (col <- 1 to query.length)
        val diag = getMatrix(row - 1, col - 1)
        val up = getMatrix(row - 1, col)
        val left = getMatrix(row, col - 1)

        if (subject(row - 1) == query(col - 1))
          setMatrix(row, col, diag + scoreInfo.matchBonus)
        else
          setMatrix(row, col, Vector(diag - scoreInfo.mismatchPenalty, up - scoreInfo.gapPenalty, left - scoreInfo.gapPenalty, 0).max)

  private enum Direction:
    case Upward, Leftward, Diagonal, Stop

  private def getDirection(row: Int, col: Int): Direction =
    val cur = getMatrix(row, col)

    val diag = getMatrix(row - 1, col - 1)
    val diagMatch = diag + scoreInfo.matchBonus
    val diagMismatch = diag + scoreInfo.mismatchPenalty

    val up = getMatrix(row - 1, col)
    val upGap = up - scoreInfo.gapPenalty

    val left = getMatrix(row, col - 1)
    val leftGap = left - scoreInfo.gapPenalty

    if cur == 0 then
      Direction.Stop
    else if diag == diagMatch && subject(row - 1) == query(col - 1) then
      Direction.Diagonal
    else if diag == diagMismatch && subject(row - 1) != query(col - 1) then
      Direction.Diagonal
    else if up == upGap then
      Direction.Upward
    else if left == leftGap then
      Direction.Leftward
    else
      Direction.Stop

  private def backtrackAlignment(): (String, String) =
    val maxScoreIndex =
      var bestV = matrix(matrix.length - 1)
      var bestI = 0

      for (i <- matrix.length - 1 to 0)
        if (matrix(i) > bestV) {
          bestV = matrix(i)
          bestI = i
        }

      bestI

    var subjectLine = StringBuilder()
    var queryLine = StringBuilder()

    var row = maxScoreIndex / colWidth
    var col = maxScoreIndex % colWidth

    def backtrackAlignmentRec(row: Int, col: Int): Unit =
      if row == 0 || col == 0 then
        return ()

      val direction = getDirection(row, col)

      direction match
        case Direction.Upward =>
          subjectLine.append(subject(row - 1))
          queryLine.append('-')
          backtrackAlignmentRec(row - 1, col)
        case Direction.Leftward =>
          subjectLine.append('-')
          queryLine.append(query(col - 1))
          backtrackAlignmentRec(row, col - 1)
        case Direction.Diagonal =>
          subjectLine.append(subject(row - 1))
          queryLine.append(query(col - 1))
          backtrackAlignmentRec(row - 1, col - 1)
        case Direction.Stop => ()

    backtrackAlignmentRec(row, col)

    (subjectLine.reverseInPlace().result(), queryLine.reverseInPlace().result())

  def invoke(): (String, String) =
    calculateMatrix()
    backtrackAlignment()
